File: bindings/python/examples/custom_function_registration.py
import formualizer as fz

wb = fz.Workbook(mode=fz.WorkbookMode.Ephemeral)
wb.add_sheet("Sheet1")

wb.set_value("Sheet1", 1, 1, 1)
wb.set_value("Sheet1", 1, 2, 2)
wb.set_value("Sheet1", 2, 1, 3)
wb.set_value("Sheet1", 2, 2, 4)


def sum_range(grid):
    return sum(cell for row in grid for cell in row)


wb.register_function("py_sum_range", sum_range, min_args=1, max_args=1)
wb.set_formula("Sheet1", 1, 3, "=PY_SUM_RANGE(A1:B2)")

print(wb.evaluate_cell("Sheet1", 1, 3))
print(wb.list_functions())

wb.unregister_function("py_sum_range")

File: bindings/wasm/examples/custom-function-registration.mjs
import init, { Workbook } from "formualizer";

await init();

const wb = new Workbook();
wb.addSheet("Sheet1");

wb.setValue("Sheet1", 1, 1, 1);
wb.setValue("Sheet1", 1, 2, 2);
wb.setValue("Sheet1", 2, 1, 3);
wb.setValue("Sheet1", 2, 2, 4);

wb.registerFunction(
  "js_sum_range",
  (grid) => grid.flat().reduce((total, value) => total + value, 0),
  { minArgs: 1, maxArgs: 1 },
);

wb.setFormula("Sheet1", 1, 3, "=JS_SUM_RANGE(A1:B2)");
wb.evaluateAll();

console.log(wb.sheet("Sheet1").getValue(1, 3));
console.log(wb.listFunctions());

wb.unregisterFunction("js_sum_range");
